File: 백준/Silver/2178. 미로 탐색/미로 탐색.js
class Queue {
  items = [];
  front = 0;
  rear = 0;

  push(item) {
    this.items.push(item);
    this.rear++;
  }

  pop() {
    return this.items[this.front++];
  }

  isEmpty() {
    return this.front === this.rear;
  }
}

function solution(input) {
  const [N, M] = input[0].split(" ").map(Number);
  const maps = input.slice(1).map((row) => row.split("").map(Number));

  const moves = [
    [-1, 0],
    [0, -1],
    [1, 0],
    [0, 1],
  ];

  const dist = Array.from({ length: N }, () => Array(M).fill(-1));

  function bfs(start) {
    const q = new Queue();
    q.push(start);
    dist[start[0]][start[1]] = 1;

    while (!q.isEmpty()) {
      const here = q.pop();

      for (const [dx, dy] of moves) {
        const row = here[0] + dx;
        const column = here[1] + dy;

        if (row < 0 || row >= N || column < 0 || column >= M) {
          continue;
        }

        if (maps[row][column] === 0) continue;

        if (dist[row][column] === -1) {
          q.push([row, column]);
          dist[row][column] = dist[here[0]][here[1]] + 1;
        }
      }
    }

    return dist;
  }

  bfs([0, 0]);

  return dist[N - 1][M - 1];
}

const input = [];
require("readline")
  .createInterface({
    input: process.stdin,
    output: process.stdout,
  })
  .on("line", (line) => {
    input.push(line);
  })
  .on("close", () => {
    console.log(solution(input));
    process.exit();
  });

File: 백준/Silver/2178. 미로 탐색/미로 탐색.py

from collections import deque
import sys
input = sys.stdin.readline

def solution():
  N, M = map(int, input().split())
  maps = [list(map(int, input().strip())) for _ in range(N)]
  
  moves = [[-1, 0], [1, 0], [0, -1], [0,1]]
  dist = [[-1]*M for _ in range(N)]
  
  def bfs(start):
    q = deque()
    
    q.append(start)
    dist[start[0]][start[1]] = 1;
    
    while q:
      x, y = q.popleft()
      
      for dx, dy in moves:
        nx = x + dx
        ny = y + dy
        
        if nx < 0 or nx >= N or ny < 0 or ny >= M:
          continue
        
        if maps[nx][ny] == 0:
          continue;
        
        if dist[nx][ny] == -1:
          q.append((nx, ny))
          dist[nx][ny] = dist[x][y] + 1
      
  bfs((0, 0))    
  print(dist[N-1][M-1])
      

solution()